### ISSUE TITLE
Where is the bug in the Toeplitz-dot-Hankel approach?

### ISSUE THREAD

**Issue by MikaelSlevinsky** [MikaelSlevinsky/FastTransforms.jl#7]
The Toeplitz-dot-Hankel approach has now been added and it's blazingly fast. But with two types of low-precomputation fast transforms, we can now run comparisons to extremely high order.

This [gist](https://gist.github.com/MikaelSlevinsky/7ad4ec0e2c44a22cc1a2199439bcf59c) uses normally distributed coefficients (seeded for reproducibility) and compares absolute error in the coefficients of the compositions of forward-inverse and inverse-forward Chebyshev--Legendre transforms.
![asycxn](https://cloud.githubusercontent.com/assets/8176037/15365613/d49a3002-1d18-11e6-8b32-6bd9b8caee4c.png)

The culprit appears to be the recently added `cheb2leg`:
```julia
julia> using FastTransforms

julia> srand(0);

julia> c = randn(100_000);

julia> norm(cjt(c,0.,0.)-leg2cheb(c),Inf)
8.992806499463768e-15

julia> norm(icjt(c,0.,0.)-cheb2leg(c),Inf)
1.2517668892542133e-6

```

**Comment by MikaelSlevinsky** [comment]
I just checked with @aaustin141 in the chebfun implementation and the CXN versions appear to have 1e-12 error at length 100_000.

Will investigate.

**Comment by ajt60gaibb** [comment]
Yes, the MATLAB implementation in Chebfun of CXN seems to far more accurate than presented here.

**Comment by MikaelSlevinsky** [comment]
if somebody could find out why, that would be amazing

**Comment by MikaelSlevinsky** [comment]
I have a test for the use of Stirling's series for a ratio of gamma functions compared with BigFloat. It appears to compute with high relative accuracy.

**Comment by MikaelSlevinsky** [comment]
It appears that the chebfun code for cheb2leg is more involved than the other cases probably for reasons of numerical stability. From what I understand,

- the right diagonal scaling is built into the columns of the Hankel matrix
- the diagonal of the Toeplitz matrix is zeroed and the diagonal of the connection matrix, with its opposite sign, is computed and applied separately

@ajt60gaibb, it would be straightforward to translate the code, but is there a way to use the existing modular, functional approach in this repository?

**Comment by dlfivefifty** [comment]
I don't remember writing cheb2leg, did someone else do this?

**Comment by MikaelSlevinsky** [comment]
I did... Kuan wanted a new tag for the package, and your paper says v0.0.4 will have the `leg2cheb`, `cheb2leg`, `jac2jac` methods via the T.*H approach. So i tried adding them as described in the paper (S 5.1 in particular for `cheb2leg`), and ended up with this issue.

**Comment by MikaelSlevinsky** [comment]
I've introduced diagonal scaling into the partial Cholesky decomposition of the Hankel part in cf090e3655e959c23556d1af681c3378ab9c3081 (thanks to @marcusdavidwebb for pointing this out!). Running the same gist now results in much better accuracy. To me, the Chebfun code has obfuscated this detail, but I think the methods are now equivalent. Are we satisfied?
![asycxn](https://cloud.githubusercontent.com/assets/8176037/18519904/314e6e4e-7a6b-11e6-9da6-4f9675b0f895.png)
